server/broadcast.go

```diff
@@ -0,0 +1,214 @@
+	"github.com/livepeer/go-livepeer/core"
+	"github.com/livepeer/go-livepeer/drivers"
+	"github.com/livepeer/go-livepeer/monitor"
+
+	"github.com/livepeer/lpms/stream"
+)
+
+func selectOrchestrator(n *core.LivepeerNode, cpl core.PlaylistManager) (*BroadcastSession, error) {
+
+	if n.OrchestratorPool == nil {
+		glog.Info("No orchestrators specified; not transcoding")
+		return nil, ErrDiscovery
+	}
+
+	rpcBcast := core.NewBroadcaster(n)
+
+	tinfos, err := n.OrchestratorPool.GetOrchestrators(1)
+	if len(tinfos) <= 0 {
+		return nil, ErrNoOrchs
+	}
+	if err != nil {
+		return nil, err
+	}
+	tinfo := tinfos[0]
+
+	// set OSes
+	var orchOS drivers.OSSession
+	if len(tinfo.Storage) > 0 {
+		orchOS = drivers.NewSession(tinfo.Storage[0])
+	}
+
+	return &BroadcastSession{
+		Broadcaster:      rpcBcast,
+		ManifestID:       cpl.ManifestID(),
+		Profiles:         BroadcastJobVideoProfiles,
+		OrchestratorInfo: tinfo,
+		OrchestratorOS:   orchOS,
+		BroadcasterOS:    cpl.GetOSSession(),
+	}, nil
+}
+
+func processSegment(cxn *rtmpConnection, seg *stream.HLSSegment) {
+
+	nonce := cxn.nonce
+	rtmpStrm := cxn.stream
+	sess := cxn.sess
+	cpl := cxn.pl
+	mid := rtmpManifestID(rtmpStrm)
+	vProfile := cxn.profile
+
+	if monitor.Enabled {
+		monitor.LogSegmentEmerged(nonce, seg.SeqNo)
+	}
+
+	seg.Name = "" // hijack seg.Name to convey the uploaded URI
+	name := fmt.Sprintf("%s/%d.ts", vProfile.Name, seg.SeqNo)
+	uri, err := cpl.GetOSSession().SaveData(name, seg.Data)
+	if err != nil {
+		glog.Errorf("Error saving segment %d: %v", seg.SeqNo, err)
+		if monitor.Enabled {
+			monitor.LogSegmentUploadFailed(nonce, seg.SeqNo, err.Error())
+		}
+		return
+	}
+	if cpl.GetOSSession().IsExternal() {
+		seg.Name = uri // hijack seg.Name to convey the uploaded URI
+	}
+	err = cpl.InsertHLSSegment(vProfile, seg.SeqNo, uri, seg.Duration)
+	if monitor.Enabled {
+		monitor.LogSourceSegmentAppeared(nonce, seg.SeqNo, string(mid), vProfile.Name)
+		glog.V(6).Infof("Appeared segment %d", seg.SeqNo)
+	}
+	if err != nil {
+		glog.Errorf("Error inserting segment %d: %v", seg.SeqNo, err)
+		if monitor.Enabled {
+			monitor.LogSegmentUploadFailed(nonce, seg.SeqNo, err.Error())
+		}
+	}
+
+	if sess == nil {
+		return
+	}
+	go func() {
+		// storage the orchestrator prefers
+		if ios := sess.OrchestratorOS; ios != nil {
+			// XXX handle case when orch expects direct upload
+			uri, err := ios.SaveData(name, seg.Data)
+			if err != nil {
+				glog.Error("Error saving segment to OS ", err)
+				if monitor.Enabled {
+					monitor.LogSegmentUploadFailed(nonce, seg.SeqNo, err.Error())
+				}
+				return
+			}
+			seg.Name = uri // hijack seg.Name to convey the uploaded URI
+		}
+
+		// send segment to the orchestrator
+		glog.V(common.DEBUG).Infof("Submitting segment %d", seg.SeqNo)
+
+		res, err := SubmitSegment(sess, seg, nonce)
+		if err != nil {
+			if shouldStopStream(err) {
+				glog.Warningf("Stopping current stream due to: %v", err)
+				rtmpStrm.Close()
```

Comment:
What is the expected behavior when an RTMP stream is closed? If we proceed to also stop the session and find a new orchestrator when is a new RTMP stream created?

Reply:
A `BroadcastSession` would still be assigned to `rtmpConnection.sess`, but should eventually be GC'd along with the `rtmpConnection`, so it's most likely harmless -- we don't have anything to manually clean up.

This is probably not the most robust solution as it could lead to a resource leak if we ever need to do a manual clean-up around `BroadcastSession` in the future.

Reply:
Improved this behavior; now we check whether the RTMP connection is still active prior to assigning `cxn.sess = sess`. If inactive, `sess` is set to nil. This should be safe since all accesses to the `rtmpConnections` map need to be protected by a mutex. See https://github.com/livepeer/go-livepeer/pull/646/commits/43d97d78b92131c41b7f66b1e5433bdb58d34d70#diff-236217e7f473dfbde9134c6a391f3408R336

Reply:
Circling back to the original question -- added a `return` after the `rtmpStrm.Close()` so we aren't doing unnecessary work after the close. https://github.com/livepeer/go-livepeer/pull/646/commits/43d97d78b92131c41b7f66b1e5433bdb58d34d70#diff-cecc27bfe2b6b19181ce443733dfba57R147